README
Forks::Super version 0.73
=========================

Forks::Super provides drop-in replacements for the Perl
fork(), wait(), and waitpid() functions with
additional features for spawning background processes
and managing them, including:

   *  forking to shell command or subroutine

      can specify a shell command or a Perl subroutine
      (name or code reference) to invoke in the child
      process

   *  timeouts for background processes

      can specify a deadline (in relative or absolute
      time) for a background process to complete,
      with the background process being killed if it does
      not complete by its deadline.

   *  throttling

      Limiting the number of simultaneous processes, or
      block new processes from starting when the system's
      CPU load is too high. Callers can install their own
      functions to determine when the system is too busy 
      to launch another job. More advanced features allow
      you to suspend and resume background tasks according
      to your own criteria.

   *  dependencies

      Jobs can be specified to wait until other jobs
      have started and/or completed before they can begin.

   *  deferred jobs

      Jobs can be specified to start at some specific time in
      the future, or to be launched under some specific
      future conditions. Arbitrary priorities can be
      assigned to jobs so that more urgent jobs will be
      launched first.

   *  simple interprocess communication

      Parent process can access a child process's standard
      input, output, and error streams to facilitate
      interprocess communication. Parent and child processes
      can share mutex objects to coordinate their activities.

   *  timeouts on wait,waitpid calls

      extended  wait  and  waitpid  functions can take
      an optional timeout argument
      
   *  operating system features

      On some operating systems, the priority and CPU affinity
      of the child processes can be set. The module also provides
      a (more) portable set of routines for signalling,
      suspending, or terminating processes.

   *  process identifier objects

      Return values from fork/wait/waitpid calls are overloaded
      objects with methods for process monitoring, signalling,
      and interprocess communication.


RECOMMENDED PERL VERSION

The Forks::Super module is recommended for versions of Perl
later than 5.7.2 that implement "deferred" signals (See the 
"Deferred Signals (Safe Signals)" section of any recent version 
of perlipc), as the signal handlers used in the module are
insufficiently paranoid. However the module will compile
and run with Perl 5.6.


INSTALLATION

The typical installation recipe

   perl Makefile.PL
   make
   make test
   make install

usually works. The unit tests assume they are
running on a moderately loaded system; intermittent
failures are more likely on a heavily loaded
system. If "make test" fails, it will often work
if you run it a second time.

It will take about 15 minutes to run the test suite
of the Forks::Super module with "make test". An
alternative is to run

   make fasttest

which will run the Forks::Super module tests in
parallel, using the Forks::Super module. Running
"make fasttest" will take about 2-3 minutes.


DEPENDENCIES

Windows users must install the  Win32::API  module.

The  bg_eval  function for evaluating Perl code in a 
background process requires either the YAML, YAML::Tiny,
JSON, or Data::Dumper modules. It is a fatal error to use  bg_eval
when at least one of these modules can not be found.

Otherwise,  Forks::Super  recommends and will make
use of the following modules if they are installed,
but will still be able to perform most of its functions
without them:

    DateTime::Format::Natural
    Proc::ProcessTable
    Win32::Process, Win32::API [Windows and Cygwin]
    Win32          [Windows]
    Sys::CpuAffinity 
    Sys::CpuLoadX

The Sys::CpuAffinity and Sys::CpuLoadX are bundled with the
Forks::Super module distribution, and may be installed at
the same time. The Sys::CpuAffinity module is also
available as a separate module on CPAN. See "BUNDLED
MODULES", below.


BUNDLED MODULES

This Forks::Super distribution comes bundled with two
additional modules, Sys::CpuAffinity and Sys::CpuLoadX. 
You will have the opportunity to install these modules
at the same time that you install the Forks::Super module.

The Sys::CpuAffinity module is for manipulating process
CPU affinities. The Forks::Super module can make use of
this module to control the CPU affinities of background
processes. This is a released module that can also be
retrieved and installed from CPAN.

The Sys::CpuLoadX module is for determining the current 
CPU load of your system. The Forks::Super module can make
use of this module to decide whether the system is too
busy to launch additional background tasks. As of Forks::Super
version 0.30, this is an unreleased module and it is only
available bundled with Forks::Super.

Installation of these additional modules is optional. If the
modules are not available, then certain features of the
Forks::Super module may not work.


WINDOWS

Some features of this module do not work or work a
little differently with Windows systems. See the
README.windows file included in this distribution
for important information.


LICENSE AND COPYRIGHT

Copyright (c) 2009-2013, Marty O'Brien.

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself, either Perl version 5.8.8 or,
at your option, any later version of Perl 5 you may have available.

See http://dev.perl.org/licenses/ for more information.


Changes
Revision history for Forks-Super

0.73    2014-06-18
        support corrective measures for too many open filehandles

0.72    2013-12-09
        Support  fork \&code  and  fork \@cmd  syntax.
        Use 'queue' as default on_busy behavior for jobs with dependencies

0.71    2013-10-10
        test fixes that should help single CPU systems pass timing tests

0.70    2013-09-18
        Experimental emulation of process groups in MSWin32 in 
        waitpid/kill functions.
        Improved procedures to terminate a process tree in MSWin32.

0.69    2013-08-19
        Move FS::Tie::BackgroundScalar and BackgroundArray, which are no
        longer used with 'tie', to FS::LazyEval.
        Better handling of recalcitrant MSWin32 processes.

0.68    2013-07-01
        Refactored sync objects - should eliminate race conditions
        during parent/child startup.

0.67    2013-05-25
        Support share attribute with a natural fork.

0.66    2013-04-16
        RT#84548 - re-fix disable overload for Perl <=5.12 too

0.65    2013-04-13
        RT#84548 - fix disable overload warning with Perl>=v5.16
        RT#76868 - incorporate Devel::GlobalDestruction if available.

0.64    2013-03-23
        Improved thread signal processing in Win32
        Upgraded bundled Sys-CpuAffinity and Signals-XSIG modules.
        Some Forks::Super::Sync improvements

0.63    2012-03-28
        refactor FS::Job::OS::get_cpu_load func for robustness, portability
        remove F::S::Util dependency in SysInfo.pm.PL for openbsd's sake

0.62    2012-03-22
        Fixed doc bug: max_fork option to F::S:fork should have been max_proc.
        Implemented configuration files.
        Implemented dynamic configuration through signals. Not sure how to
        document it yet.

0.61    2012-03-05
        Defines constants BG_QX, BG_EVAL as shorthand for tied class names.
        Miscellaneous fixes specific to freebsd, openbsd, and solaris.

0.60    2012-02-29
        tie interface to  bg_qx, bg_eval  calls. These calls are equivalent
        in function if not in implementation:
             $foo = bg_qx $command, \%options
             tie $foo, 'Forks::Super::bg_qx', $command, \%options
        And now supports the clunky list context background evals
             tie @output, 'Forks::Super::bg_qx', $command, \%options
             tie %result, 'Forks::Super::bg_eval', sub { ... }, \%options
        A few more tweaks that make the module friendlier to Perl v<=5.6

0.59    2012-02-22
        Improved waitall performance where there are many jobs to reap.
        Improved error handling in sync option Win32 implementations. 

0.58    2011-12-15
        Added Java-style dump of program information in response to SIGQUIT
        or other signal. See  Forks::Super::Debug::enable_dump.
        Minor fixes and test tweaks for solaris and openbsd

0.57    2011-12-09
        Added sync =>  option for mutexs to coordinate parent and child procs.
        New getc_XXX methods for interprocess communication.
        Test fixes and tweaks.

0.56    2011-11-21
        Fix build process bug introduced in v0.55. Failure to build bundled
        Sys-CpuAffinity distro now won't break the whole build.

0.55    2011-11-10
        exec and timeout are now compatible options
        timeout  now suppored on systems that don't implement  alarm , or
        where  sleep  and  alarm  are incompatible.
        Test for -MForks::Super=cleanse
        block => 1  option now respected with child's STDIN.
        Upgrade bundled Sys::CpuAffinity to v1.05
        Create Forks::Super::Debug::parent_dump for JVM style job dumps

0.54    2011-10-07
        Allows PerlIO layers in  child_fh  spec.
        Upgraded bundled Signals-XSIG to v0.13, Sys-CpuAffinity to 1.04.
        More effort backporting to Perl v5.6.
        kill func can signal procs that were not launched by Forks::Super.
        Improvements to suspend/resume functions in Cygwin.
        Fixes for daemon code introduced in v0.53.
        Perl::Critic-inspired refactoring.

0.53    2011-07-14
        Introduce  daemon  option to fork.
        Introduce  share => list-of-refs  option to fork.
        Introduce  env => \%kvpairs , umask => mask  options to fork.
        Fixed -MForks::Super=cleanse bug introduced in 0.52
        Forks::Super::kill on cmd-style job now delivers signal to the external
        command, not the perl process that called system().
        Increased initial deferred job id from -100K to -500K.
        MSWin32 refactoring: fork-to-cmd, fork-to-exec, OS priority code.
        bg_eval, bg_qx commands now return empty string on failure, never undef.
        Upgrade Signals::XSIG bundle to v0.12.

0.52    2011-06-21
        Calling Forks::Super::LazyEval::bg_eval and bg_qx in list context
        doesn't work and is now disabled.
        Performance enhancement for fork with file-based IPC.
        Upgrade Sys-CpuAffinity bundle to v1.01
        Bug fix in timeout code that hangs BSD (looks like it didn't work)
        Test fixes.

0.51    2011-04-18
        Include IPCPipeHandle.pm in manifest (D'oh)
        Overload <> operator for Forks::Super::Job to read from child stdout.
        $job->wait(0) now waits with timeout of 0, behaves differently
        from $job->wait().

0.50    2011-04-17
        Support  dir => $directory  option to start child in arbitrary dir.
        Introduced tied F::S::Tie::IPCPipeHandle class for testing/debugging
        use of pipes.
        Support  cpu_affinty => [@list]  syntax.
        Calling exit from a child sub is allowed again.

0.49    2011-04-01
        Bundle with Signals::XSIG v0.10 and drop Forks::Super::Sighandler.
        Setting $ENV{IPC_DIR}="undef" to disable file-based IPC now functions
        except on MSWin32.

0.48    2011-03-16
        Fix compile error in t/49.
        Refactoring of some large subroutines.

0.47    2011-03-15
        Fix for IPC clean up. Removed obsolete files and tests.

Makefile.PL
use strict;
use warnings;
use ExtUtils::MakeMaker;
use v5.006; # 5.7.1 or better recommended

our %RECOMMENDED
    = ('YAML' => '0.62',
       'JSON' => '2.0',
       'YAML::Tiny' => '1.39',
       'Sys::CpuAffinity' => '0.91',
       'Sys::CpuLoadX' => '0.02',
       'DateTime::Format::Natural' => '0.89',
       'PerlIO::gzip' => 0 );

if ($^O eq 'cygwin') {
    $RECOMMENDED{'Win32::API'} = '0.71';
}
if ($^O eq 'MSWin32') {
    %RECOMMENDED = (%RECOMMENDED,
		    'Win32::API' => '0.71',
		    'Win32::IPC' => 0,
		    'Win32::Process' => '0.14',
		    'Win32::Process::Kill' => '2.64',
		    'DBD::WMI' => 0);
}

our %REQUIRED 
    = ('Test::More' => '0',
       'Time::HiRes' => '1.30');
if ($^O eq 'MSWin32') {
    $REQUIRED{'Win32::API'} = '0.43';
    # Win32::Process? 
}

# files that may exist and be nice to clean up with a  make clean
our $FILES_TO_CLEAN 
    = join " ", qw(Forks-Super-* system-limits system-limits.lock
		   lib/Forks/Super/SysInfo.pm* t/out/spawn* t/out/07*
		   t/out/sigchld.debug t/out/test* t/out/.lock-*
		   t/out/48b.* t/out/debug1-* t/out/limits.* t/poorsort.pl
		   t/out/ipc* t/out/new-ipc* t/out/66* t/dir* t/out/24*
                   t/out/daemon* lib/_Inline *.stackdump t/out/.sync*);

our %META_MERGE
    = (recommends => \%RECOMMENDED,
       no_index => { 
	   'package' => ['Sys::CpuAffinity', 
			 'Signals::XSIG',
			 'Signals::XSIG::Default',
	       	         'Signals::XSIG::TieArray56'] });

######################################################################

if ($] < 5.007003) {

    # common but intermittent test failures on linux v5.6.2:
    #   t/32
    #   t/30b
    #   t/48a
    #   t/43g
    #   t/56a
    #   t/44g
    #   and others

    warn qq{

WARNING !!! The Forks::Super module makes heavy use of 
insufficiently paranoid signal handlers (See "Deferred
Signals (Safe Signals)" in a more recent version of perlipc) 
and is not recommended on versions of perl older than
v5.7.3. (Your version: $])

};
    sleep 1;
}


# Win32::API v0.71 strongly recommended on Windows
if ($^O eq 'MSWin32') {
    my $V = eval 'use Win32::API; $Win32::API::VERSION';
    if ($@) {
	warn qq[

Win32::API not found. That module is strongly recommended
with this distribution on Windows systems (at least v0.71).

];
    } elsif ($V < 0.71) {
	warn qq[

Win32::API v$V found. This distribution strongly recommends
that you upgrade Win32::API to at least v0.71.

];
    }
}

if (" @ARGV " =~ / bundle\b/i || " @ARGV " =~ / reinstall\b/i) {
    if ("@ARGV" =~ /bundle=0/i) {
	$ENV{BUNDLE} = '0';
    } else {
	$ENV{BUNDLE_REINSTALL} = 1;
    }
    @ARGV = grep { !/^bundle\b/i && !/^reinstall\b/ } @ARGV;
}
if (defined($ENV{BUNDLE}) && $ENV{BUNDLE} eq "0") {
    print STDERR "Skip check and install for bundled modules.\n";
} else {
    install_bundled_modules();
}
probe_system();

my $MM_VERSION = eval $ExtUtils::MakeMaker::VERSION;

WriteMakefile(

    NAME                => 'Forks::Super',
    AUTHOR              => q[Marty O'Brien <mob\@cpan.org>],       #'],
    VERSION_FROM        => 'lib/Forks/Super.pm',
    ABSTRACT_FROM       => 'lib/Forks/Super.pm',
    PL_FILES            => {},
    PREREQ_PM => \%REQUIRED,

    ($MM_VERSION >= 6.46   ? ('META_MERGE' => \%META_MERGE ) : ()),
    ($MM_VERSION >= 6.3002 ? ('LICENSE'    => 'perl')        : ()),

    dist                => { COMPRESS => 'gzip -9f', SUFFIX => 'gz', },
    clean               => { FILES => $FILES_TO_CLEAN },

);


print STDERR q[

"Makefile.PL" complete. After you run "make", you can run "make test"
to run the Forks::Super unit tests in series, or "make fasttest" to use
the Forks::Super framework to run the tests in parallel.


];

#############################################################################
#
# run installer for the additional modules
# that were included with this distribution.
#
# run  perl Makefile.PL BUNDLE=1
# or   BUNDLE_REINSTALL=1 perl Makefile.PL
# to allow re-installation of bundled modules
#
sub install_bundled_modules {
    if (-d 'bundle') {
	chdir 'bundle';
	opendir my $sub_dh, '.';
	while (my $submod = readdir $sub_dh) {
	    next if $submod eq '.';
	    next if $submod eq '..';
	    next if ! -d $submod;

	    if (-f "$submod/bundle.pl") {
		print "--------------------------------------------------\n";
		print "Running builder for submodules/$submod\n";
		chdir $submod;
		system($^X,'bundle.pl',@ARGV);
		chdir '..';
	    }
	}
	print "--------------------------------------------------\n";
	closedir $sub_dh;
	chdir '..';
    }
}

sub probe_system {
    unless ($ENV{NOLIMITS}) {
	print STDERR "Gathering information on current system $^O.$]\n";
	print STDERR "and creating Forks::Super::SysInfo package.\n";
	system($^X, 'SysInfo.pm.PL', 'lib/Forks/Super/SysInfo.pm');
    }
}


#############################################################################

package MY;

# # # # # # # # # # # # # # # # # # # # # # # # # # # # # # 
# additional testing targets
# # # # # # # # # # # # # # # # # # # # # # # # # # # # # # 

sub MY::postamble {

    my $stresstestargs = "-r 20 -x 5 -s -q -t 150";
    if ($^O =~ /solaris/i) {
	$stresstestargs = "-r 25 -x 4 -s -q -t 150";
    }
    # $(PERLRUN) not avail in 5.6.1, ExtUtils::MakeMaker 5.45
    my $PERLEXE = '$(FULLPERL)';  
    my $postamble = <<"__END_POSTAMBLE__";

# # # # # # # # # # # # # # # # # # # # # # # # # # # # # # 
#
# Additional testing targets for Forks::Super module.
#
# # # # # # # # # # # # # # # # # # # # # # # # # # # # # # 

# ------ fasttest: use Forks::Super to run Forks::Super tests in parallel

fasttest :: pure_all
	$PERLEXE t/forked_harness.pl \$(TEST_FILES) -h -q -t 150

# ------ stress test: run all tests in parallel 100 times

stresstest :: pure_all
	$PERLEXE t/forked_harness.pl \$(TEST_FILES) $stresstestargs

__END_POSTAMBLE__

    return $postamble;
}
